flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java

```diff
@@ -60,318 +69,470 @@
-            List<RexNode> projects = ((Project) rel).getProjects();
-            ImmutableIntList inputJoinKeys = getInputIndices(projects, joinKeys);
-            if (inputJoinKeys.isEmpty()) {
-                factSideFactors.isSuitableJoinKey = false;
-                return;
+        /**
+         * Not all filter condition suitable for using to filter partitions by dynamic partition
+         * pruning rules. For example, NOT NULL can only filter one default partition which have a
+         * small impact on filtering data.
+         */
+        private static boolean isSuitableFilter(RexNode filterCondition) {
+            switch (filterCondition.getKind()) {
+                case AND:
+                    List<RexNode> conjunctions = RelOptUtil.conjunctions(filterCondition);
+                    return isSuitableFilter(conjunctions.get(0))
+                            || isSuitableFilter(conjunctions.get(1));
+                case OR:
+                    List<RexNode> disjunctions = RelOptUtil.disjunctions(filterCondition);
+                    return isSuitableFilter(disjunctions.get(0))
+                            && isSuitableFilter(disjunctions.get(1));
+                case NOT:
+                    return isSuitableFilter(((RexCall) filterCondition).operands.get(0));
+                case EQUALS:
+                case GREATER_THAN:
+                case GREATER_THAN_OR_EQUAL:
+                case LESS_THAN:
+                case LESS_THAN_OR_EQUAL:
+                case NOT_EQUALS:
+                case IN:
+                case LIKE:
+                case CONTAINS:
+                case SEARCH:
+                case IS_FALSE:
+                case IS_NOT_FALSE:
+                case IS_NOT_TRUE:
+                case IS_TRUE:
+                    // TODO adding more suitable filters which can filter enough partitions after
+                    // using this filter in dynamic partition pruning.
+                    return true;
+                default:
+                    return false;
             }
+        }
 
-            visitFactSide(rel.getInput(0), factSideFactors, inputJoinKeys);
-        } else if (rel instanceof Calc) {
-            Calc calc = (Calc) rel;
-            RexProgram program = calc.getProgram();
-            List<RexNode> projects =
-                    program.getProjectList().stream()
-                            .map(program::expandLocalRef)
-                            .collect(Collectors.toList());
-            ImmutableIntList inputJoinKeys = getInputIndices(projects, joinKeys);
-            if (inputJoinKeys.isEmpty()) {
-                factSideFactors.isSuitableJoinKey = false;
-                return;
+        private void setTables(ContextResolvedTable catalogTable) {
+            if (tables.size() == 0) {
+                tables.add(catalogTable);
+            } else {
+                for (ContextResolvedTable thisTable : new ArrayList<>(tables)) {
```

Comment:
nit: why we need new ArrayList instance here?

Reply:
> nit: why we need new ArrayList instance here?

The list variable `tables` have an `add` operation in the for loop. If we don't new one new ArrayList, a concurrent modification exception will be thrown.